crates/shared/src/token_list.rs

```diff
@@ -1,12 +1,43 @@
-use std::collections::HashMap;
+use std::{collections::HashMap, sync::Mutex};
 
 use anyhow::Result;
 use ethcontract::H160;
-use reqwest::{Client, IntoUrl};
+use reqwest::Client;
 use serde::Deserialize;
 
+use crate::maintenance::Maintaining;
+
+#[derive(Clone, Debug, Default)]
+pub struct TokenListConfiguration {
+    pub url: String,
+    pub chain_id: u64,
+    pub client: Client,
+}
+
+impl TokenListConfiguration {
+    async fn tokens(&self) -> Result<HashMap<H160, Token>> {
+        let model: TokenListModel = self
+            .client
+            .get(self.url.clone())
+            .send()
+            .await?
+            .json()
+            .await?;
+        Ok(Self::from_tokens(model.tokens, self.chain_id))
+    }
+
+    fn from_tokens(tokens: Vec<TokenModel>, chain_id: u64) -> HashMap<H160, Token> {
+        tokens
+            .into_iter()
+            .filter(|token| token.chain_id == chain_id)
+            .map(|token| (token.token.address, token.token))
+            .collect()
+    }
+}
+
 pub struct TokenList {
-    tokens: HashMap<H160, Token>,
+    tokens: Mutex<HashMap<H160, Token>>,
```

Comment:
I think it's worth mentioning that a [`DashMap`](https://lib.rs/crates/dashmap) is typically vastly better than a `Mutex<HashMap>`. However it appears to me that this struct won't experience much concurrent access, so `Mutex<HashMap>` is probably good enough.

Reply:
Didn't know about it, but if we would introduce new crate it would be nice to do refactor in a separate PR and change all occurrences in the code.